semgrep/semgrep/output.py

```diff
@@ -156,6 +156,8 @@ def build_sarif_output(rule_matches: List[RuleMatch], rules: FrozenSet[Rule]) ->
     - which links to this schema https://github.com/oasis-tcs/sarif-spec/blob/master/Schemata/sarif-schema-2.1.0.json
     - full spec is at https://docs.oasis-open.org/sarif/sarif/v2.1.0/cs01/sarif-v2.1.0-cs01.html
     """
+    # SARIF doesn't allow duplicate IDs, we dedupe here
+    rules_by_id = {rule.id: rule for rule in rules}
```

Comment:
I think it'd be worth it to keep digging and see why we're hitting this issue. If there's a bug earlier in the program causing this then we'd like to know about it.

What I think is happening: we currently have nothing preventing multiple rules from using the same rule ID. There's been some discussion around this recently, but I think this is undesirable behavior. IMO we should fail to run if `semgrep` detects multiple rules with the same ID. In this case I suspect we have a couple rules using the same ID, which is causing this issue.

Not only is this undesirable from a rule cleanliness point of view, but I believe these changes will introduce other subtle bugs. E.g. with these changes, if two rules use the same ID we will silently ignore some findings:

```
In [1]: {r['id']: r['value'] for r in [{'id': 1, 'value': 'foo'}, {'id': 1, 'value': 'bar'}]}           
Out[1]: {1: 'bar'}
```

Even worse, we will only do this with SARIF output, so other outputs will appear normal. I think the fix here is to revamp our rule IDs. Early in the program we should detect duplicate IDs and warn the user and error out, IMO.

Reply:
That's quite possible. Are we certain that multiple rules with the same rule ID is a bug and not a feature, though?

Reply:
> That's quite possible. Are we certain that multiple rules with the same rule ID is a bug and not a feature, though?

I definitely think it's a bug, but I'm open to hearing why it may be a feature. Here are some problems I see with duplicate rule IDs:

* This breaks rule discovery/usage in a number of different contexts: Google search, Github issues,`nosem` functionality, any future features around whitelisting/blacklisting individual rules, etc. I.e. using a rule ID in any of those contexts no longer lands me on a single rule. Google searching a rule ID may bring up two different results for two different people. Using `nosem` exceptions may disable more rules than intended. Rules should be one-to-one with their rule ID, at least insofar as we can control.
* As seen here, most tools assume a rule or linter is one-to-one with its ID. As far as I know every other linter/static analysis tool maintains this one-to-one relationship. Breaking this assumption will make it difficult to play well with other tools or systems.
* Also as seen here, a deficiency with duplicate rule IDs is the tendency to use `dict`s from ID -> rule, which will silently remove rules with the same ID from consideration. This is mostly something we'll have to consider internally, but a design prone to subtle bugs is not ideal.

I'd be curious to hear the advantages to duplicate rule IDs and how they weigh against the costs.

Reply:
> I'm open to hearing why it may be a feature

I have no idea, I couldn't really understand what's going on in the Rule class. I was imagining stuff like, maybe a user would want to use multiple rule definitions with the same ID but various path include/excludes. One such case that comes to mind now is:

```yaml
rules:
  - id: wrong-user-agent
     patterns:
       - pattern: user_agent="..."
       - pattern-not: user_agent="foo"
     paths:
        - src/foo/**
  - id: wrong-user-agent
     patterns:
       - pattern: user_agent="..."
       - pattern-not: user_agent="bar"
     paths:
        - src/bar/**
```

Reply:
Big 👍  to @mschwager comments about the downsides of duplicate IDs. I'd lean towards not letting a user define the same `id` in a single YAML file -- more trouble than it's worth to diagnose later, and a very uncommon usecase if we did make it a feature IMO.